File: Challenge/May_LeetCoding_Challenge/Week_2:_April_8th–April_14th/208_Implement_Trie_(Prefix_Tree)/Trie.java
class Trie {
    final private Trie[] children = new Trie[26];
    private boolean isLeaf;

    /** Initialize your data structure here. */
    public Trie() {
    }

    /** Inserts a word into the trie. */
    public void insert(String word) {
        insert(word, 0);
    }

    private void insert(String word, int idx) {
        if (idx < word.length()) {
            int childIdx = word.charAt(idx) - 'a';
            if (children[childIdx] == null) {
                children[childIdx] = new Trie();
            }
            children[childIdx].insert(word, idx + 1);
        } else {
            isLeaf = true;
        }
    }

    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        return search(word, 0, true);
    }

    private boolean search(String word, int idx, boolean shouldMatchAll) {
        if (idx < word.length()) {
            int childIdx = word.charAt(idx) - 'a';
            return children[childIdx] != null && children[childIdx].search(word, idx + 1, shouldMatchAll);
        } else {
            return !shouldMatchAll || isLeaf;
        }
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        return search(prefix, 0, false);
    }
}

File: Algorithms/208_Implement_Trie_(Prefix_Tree)/trie.go
package main

type Trie struct {
    isLeaf   bool
    children [26]*Trie
}

/* Initialize your data structure here. */
func Constructor() Trie {
    return Trie{}
}

func (trie *Trie) insert(word string, idx int) {
    if idx < len(word) {
        childIdx := int(word[idx] - 'a')
        if trie.children[childIdx] == nil {
            trie.children[childIdx] = &Trie{}
        }
        trie.children[childIdx].insert(word, idx+1)
    } else {
        trie.isLeaf = true
    }
}

/* Inserts a word into the trie. */
func (trie *Trie) Insert(word string) {
    trie.insert(word, 0)
}

func (trie *Trie) search(word string, idx int, shouldMatchAll bool) bool {
    if idx < len(word) {
        childIdx := int(word[idx] - 'a')
        return trie.children[childIdx] != nil && trie.children[childIdx].search(word, idx+1, shouldMatchAll)
    } else {
        return !shouldMatchAll || trie.isLeaf
    }
}

/* Returns if the word is in the trie. */
func (trie *Trie) Search(word string) bool {
    return trie.search(word, 0, true)
}

/* Returns if there is any word in the trie that starts with the given prefix. */
func (trie *Trie) StartsWith(prefix string) bool {
    return trie.search(prefix, 0, false)
}
